documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+{
+    public ref readonly GamepadState this[int index] { get; }
+    GamepadState IReadOnlyList<GamepadState>.this[int index] { get; }
+    
+    public <implementation defined> GetEnumerator();
+    IEnumerator<GamepadState> IEnumerable<GamepadState>.GetEnumerator();
+    
+    public event Action<ButtonEvent<GamepadState, JoystickButton>>? ButtonDown;
+    public event Action<ButtonEvent<GamepadState, JoystickButton>>? ButtonUp;
+    public event Action<AxisEvent<GamepadState, Vector2>>? ThumbstickMove;
+    public event Action<AxisEvent<GamepadState, float>>? TriggerMove;
+}
+
+public partial class Joysticks : IReadOnlyList<JoystickState>
+{
+    public ref readonly JoystickState this[int index] { get; }
+    JoystickState IReadOnlyList<JoystickState>.this[int index] { get; }
+    
+    public <implementation defined> GetEnumerator();
+    IEnumerator<JoystickState> IEnumerable<JoystickState>.GetEnumerator();
+
+    public event Action<ButtonEvent<JoystickState, JoystickButton>>? ButtonDown;
+    public event Action<ButtonEvent<JoystickState, JoystickButton>>? ButtonUp;
+    public event Action<AxisEvent<JoystickState, float>>? AxisMove;
+    public event Action<AxisEvent<JoystickState, Vector2>>? HatMove;
+}
+```
+
+Unlike 1.0 and 2.0, this proposal uses `readonly record struct`s as their only argument for the event action. This allows us to provide more information to the event handlers without breaking in the future. These types are farily simple:
+
+```cs
+public readonly record struct ButtonEvent<TDevice, TButton>(in TDevice DeviceState, TButton Button);
+public readonly record struct ClickEvent<TDevice, TButton>(in TDevice DeviceState, TButton Button, Vector2 Position);
+public readonly record struct AxisEvent<TDevice, TAxis>(in TDevice DeviceState, int Index, TAxis OldValue, TAxis NewValue);
+public readonly record struct TextEvent<TDevice>(in TDevice DeviceState, string? OldText, string? NewText);
+```
+
+This is the part of this proposal that incorporates the ideas in Enhanced Input Events, and is why this proposal supersedes that one.
+
+One final point to note is that throughout the rest of the proposal the following type will be used:
+
+```cs
+public struct InputReadOnlyList<T> : IReadOnlyList<T>
+{
+    public InputReadOnlyList(IReadOnlyList<T> other);
+}
+```
+
+The Silk.NET team wishes to reserve the right to add more constructors to this type as it sees fit.
+
+This exists so that, should the Silk.NET choose to, we can optimize the lookup of elements while ensuring things like indexers are inlined and don't result in a virtual call if our implementation allows us to do so.
+
+**INFORMATIVE TEXT:** For example, for joystick and mouse buttons we could use a fixed-sized bit buffer where each bit represents an individual button: 1 for pressed, 0 for unpressed. But for something like keyboard input where there are a large amount of keys, we can't do that and will likely use `Memory<T>` instead.
```

Comment:
I don't understand this section, what is it for? Why do we define this struct here and how does it help with optimizing many different scenarios?

Reply:
This particular line is informative text, so it doesn't form part of the proposal at all - rather just describes the author's intent. We are completely free to ignore this line, and this is just giving _examples_ of the sorts of optimizations we could do.

I think the comment speaks for itself, and it's unobjectionable that an aggressively-inlined, non-virtual call that just does bitwise arithmetic is much more efficient than a virtual call iterating through a collection.

Reply:
Right my problem is that this section suggests that we may use *different* implementations based on type or whatever, but this is a struct, there is no way to have multiple different implementations. There is no way to have both fixed-size bit buffer for mouse and a Memory<T> for keyboards.

Reply:
Well this doesn't suggest anything as it's **informative text**, all the proposal cares about is that it's a list, it's a struct, and is constructible from another list.

We have reserved the right to make more constructors above, which will allow us to make fast paths available to external backends.

It is my intention for this type to be constructible from any `IReadOnlyList<T>`, hence why I have specified that constructor here. 

**INFORMATIVE TEXT:** Yeah, we'll likely be making a size trade-off for the fast path as you're right we can't have different implementations/memory layout per T.

This is an implementation detail that shouldn't impact the proposal - as far as the proposal is concerned there is no fast path, slow path, etc.

Reply:
Hmm, I'm not necessarily opposed to having this, just very confused why it's necessary. If we can't actually make a good fast path having this implementation detail doesn't seem logical.
Also, I think it is now valid to just return another implementing type when implementing an interface, so I think we could define all interfaces to be IReadOnlyList<T> and then where implementing we could just return an explicit, better, type?

Reply:
We can make a good fast path, but we can't switch one out per path without losing the actual fastness of the path due to a virtual call. As such, this just exists to be a thin wrapper over whatever we use. Also worth noting `Memory<T>` doesn't implement `IReadOnlyList<T>` so we'd need to do this somehow anyway.

But obviously, this is all informative text so it all doesn't matter and you can yell at me all you want when it comes to implementing it. The purpose of this section of the proposal is just to make sure you're happy with us having a struct list type so we can have optimizations that don't necessarily box and can be aggressively inlined.